src/Compilers/CSharp/Test/Emit3/Semantics/CollectionExpressionTests_WithElement_Constructor.cs

```diff
@@ -0,0 +1,1365 @@
+                public MyList(__arglist) : base() { }
+            }
+            
+            class C
+            {
+                void M()
+                {
+                    MyList list = [with(__arglist(10, "test"))];
+                }
+            }
+            """;
+
+        CreateCompilation(source).VerifyDiagnostics();
+    }
+
+    [ConditionalFact(typeof(WindowsOnly))]
+    public void WithElement_ArgList_Empty()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class MyList : List<int>
+            {
+                public MyList(__arglist) : base() 
+                {
+                    Console.WriteLine("ArgList constructor called");
+                }
+            }
+            
+            class C
+            {
+                static void Main()
+                {
+                    MyList list = [with(__arglist()), 1];
+                    Console.WriteLine(list.Count);
+                }
+            }
+            """;
+
+        CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("ArgList constructor called\r\n1"));
+    }
+
+    #endregion
+
+    #region Constructor Overload Resolution Tests
+
+    [Fact]
+    public void WithElement_OverloadResolution_ExactMatch()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                public string ConstructorUsed { get; }
+                
+                public MyList(int capacity) : base(capacity)
+                {
+                    ConstructorUsed = "int";
+                }
+                
+                public MyList(long capacity) : base((int)capacity)
+                {
+                    ConstructorUsed = "long";
+                }
+            }
+            
+            class C
+            {
+                static void Main()
+                {
+                    MyList<int> list1 = [with(10), 1];
+                    MyList<int> list2 = [with(10L), 2];
+                    Console.WriteLine($"{list1.ConstructorUsed},{list2.ConstructorUsed}");
+                }
+            }
+            """;
+
+        CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("int,long"));
+    }
+
+    [Fact]
+    public void WithElement_OverloadResolution_Ambiguous()
+    {
+        var source = """
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                public MyList(int value) : base() { }
+                public MyList(long value) : base() { }
+            }
+            
+            class C
+            {
+                void M()
+                {
+                    short s = 10;
+                    MyList<int> list = [with(s)];
+                }
+            }
+            """;
+
+        CreateCompilation(source).VerifyDiagnostics();
```

Comment:
Please verify output and demonstrate which constructor is chosen.

Reply:
5a79e3f6550